File: ios/Sources/TaigiKeyboard/App/Tabs/Tab3/Models/DictionarySearchViewModel.swift
import Foundation

/// ViewModel for dictionary search in Tab 3
@MainActor
final class DictionarySearchViewModel: ObservableObject {
    @Published var searchText = ""
    @Published var results: [DictionarySearchResult] = []
    @Published var isSearching = false

    private var searchTask: Task<Void, Never>?
    private let repository = DictionaryRepository.shared
    private let logger = DebugLogger(category: "DictionarySearchVM")

    init() {
        // Trigger LexiconService init to ensure Trie loads
        _ = LexiconService.shared
    }

    /// Build set of enabled dictionary sources from current settings
    private static func enabledSources() -> Set<DictionarySource> {
        let s = SharedSettings.shared
        var sources: Set<DictionarySource> = [.dev, .custom]
        if s.moeDictEnabled { sources.insert(.kautian) }
        if s.newwordDictEnabled { sources.insert(.taigitv) }
        if s.kunggeDictEnabled { sources.insert(.kungge) }
        if s.iTaigiDictEnabled { sources.insert(.itaigi) }
        if s.taiwanJapanDictEnabled { sources.insert(.taijit) }
        if s.taiHuaDictEnabled { sources.insert(.taihoa) }
        if s.taiwanPlantDictEnabled { sources.insert(.sitbut) }
        if s.sttiDictEnabled { sources.insert(.stti) }
        if s.khpooDictEnabled { sources.insert(.khpoo) }
        if s.khiin { sources.insert(.khiin) }
        if s.lkkDictEnabled { sources.insert(.lkk) }
        return sources
    }

    /// Called when search text changes; debounces 300ms then searches
    func onSearchTextChanged() {
        searchTask?.cancel()

        let query = searchText.trimmingCharacters(in: .whitespaces)

        guard !query.isEmpty else {
            results = []
            isSearching = false
            return
        }

        isSearching = true

        searchTask = Task {
            // 300ms debounce
            try? await Task.sleep(nanoseconds: 300_000_000)
            guard !Task.isCancelled else { return }

            do {
                let inputMode = SharedSettings.shared.inputMode

                // Detect CJK input and use hanzi search path
                let isCJK = query.unicodeScalars.contains {
                    (0x4E00 ... 0x9FFF).contains($0.value) ||
                        (0x3400 ... 0x4DBF).contains($0.value) ||
                        (0x20000 ... 0x2A6DF).contains($0.value)
                }

                logger.debug("[SEARCH] query='\(query)' isCJK=\(isCJK) inputMode=\(String(describing: inputMode))")

                let searchResults: [DictionarySearchResult]
                if isCJK {
                    searchResults = try await repository.searchByHanzi(
                        query: query,
                        inputMode: inputMode,
                        limit: 20,
                    )
                    logger.debug("[SEARCH] hanzi path returned \(searchResults.count) results")
                } else {
                    searchResults = try await repository.searchWithSources(
                        input: query,
                        inputMode: inputMode,
                        limit: 20,
                    )
                    logger.debug("[SEARCH] roman path returned \(searchResults.count) results")
                }

                // Also search custom dictionary (only for romanization input)
                let customResults: [DictionarySearchResult]
                if isCJK {
                    customResults = []
                } else {
                    let isToneAware = query.contains { $0.isNumber }
                    let searchPrefix = isToneAware
                        ? query.lowercased()
                        .replacingOccurrences(of: "-", with: "")
                        .replacingOccurrences(of: " ", with: "")
                        : CustomDictionaryService.generateNotone(query)
                    let customEntries = CustomDictionaryRepository.shared.searchSync(
                        prefix: searchPrefix,
                        isToneAware: isToneAware,
                        limit: 20,
                    )
                    customResults = customEntries.map { entry in
                        DictionarySearchResult(
                            id: -2,
                            roman: entry.roman,
                            tl: entry.roman,
                            hanzi: entry.hanzi,
                            frequency: Int.max,
                            sources: [.custom],
                        )
                    }
                }

                guard !Task.isCancelled else { return }
                // Sort: kautian (教育部) first, then by frequency
                let sorted = searchResults.sorted { a, b in
                    let aIsMoe = a.sources.contains(.kautian)
                    let bIsMoe = b.sources.contains(.kautian)
                    if aIsMoe != bIsMoe { return aIsMoe }
                    return a.frequency > b.frequency
                }
                // Filter source tags to only show enabled dictionaries
                let enabledSources = Self.enabledSources()
                let filtered = sorted.map { result in
                    DictionarySearchResult(
                        id: result.id,
                        roman: result.roman,
                        tl: result.tl,
                        hanzi: result.hanzi,
                        frequency: result.frequency,
                        sources: result.sources.filter { enabledSources.contains($0) },
                    )
                }
                results = customResults + filtered
                isSearching = false
            } catch {
                guard !Task.isCancelled else { return }
                logger.error("[SEARCH] Failed: \(error.localizedDescription)")
                results = []
                isSearching = false
            }
        }
    }
}

File: android/app/src/main/java/com/siansiansu/taigikeyboard/ui/tabs/tab3/DictionarySearchViewModel.kt
package com.siansiansu.taigikeyboard.ui.tabs.tab3

import android.app.Application
import android.util.Log
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.siansiansu.taigikeyboard.BuildConfig
import com.siansiansu.taigikeyboard.ime.core.CompositionRoot
import com.siansiansu.taigikeyboard.ime.core.PrefHelper
import com.siansiansu.taigikeyboard.ime.dictionary.CustomDictionaryDerivation
import com.siansiansu.taigikeyboard.ime.dictionary.DictionarySearchResult
import com.siansiansu.taigikeyboard.ime.dictionary.DictionarySource
import com.siansiansu.taigikeyboard.ime.dictionary.ToneConverterModels
import kotlinx.coroutines.FlowPreview
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.launch

// ViewModel for dictionary search in Tab 3 (詞庫)
class DictionarySearchViewModel(
    application: Application,
) : AndroidViewModel(application) {
    companion object {
        private const val TAG = "DictionarySearchVM"
        private const val SEARCH_DEBOUNCE_MILLIS = 300L
        private const val SEARCH_RESULT_LIMIT = 20
    }

    private val _searchText = MutableStateFlow("")
    val searchText: StateFlow<String> = _searchText.asStateFlow()

    private val _results = MutableStateFlow<List<DictionarySearchResult>>(emptyList())
    val results: StateFlow<List<DictionarySearchResult>> = _results.asStateFlow()

    private val _isSearching = MutableStateFlow(false)
    val isSearching: StateFlow<Boolean> = _isSearching.asStateFlow()

    private val prefs = PrefHelper(application)
    private val root = CompositionRoot.shared(application)

    init {
        observeSearchText()
    }

    // Build set of enabled dictionary sources from current preferences
    private fun buildEnabledSources(): Set<DictionarySource> {
        val sources = mutableSetOf(DictionarySource.DEV, DictionarySource.CUSTOM)
        if (prefs.moeDictEnabled) sources.add(DictionarySource.KAUTIAN)
        if (prefs.newwordDictEnabled) sources.add(DictionarySource.TAIGITV)
        if (prefs.kunggeDictEnabled) sources.add(DictionarySource.KUNGGE)
        if (prefs.itaigiDictEnabled) sources.add(DictionarySource.ITAIGI)
        if (prefs.taiwanJapanDictEnabled) sources.add(DictionarySource.TAIJIT)
        if (prefs.taiHuaDictEnabled) sources.add(DictionarySource.TAIHOA)
        if (prefs.taiwanPlantDictEnabled) sources.add(DictionarySource.SITBUT)
        if (prefs.sttiDictEnabled) sources.add(DictionarySource.STTI)
        if (prefs.khpooDictEnabled) sources.add(DictionarySource.KHPOO)
        if (prefs.khiin) sources.add(DictionarySource.KHIIN)
        if (prefs.lkkDictEnabled) sources.add(DictionarySource.LKK)
        return sources
    }

    fun updateSearchText(text: String) {
        _searchText.value = text
    }

    @OptIn(FlowPreview::class)
    private fun observeSearchText() {
        viewModelScope.launch {
            _searchText
                .debounce(SEARCH_DEBOUNCE_MILLIS)
                .distinctUntilChanged()
                .collect { query ->
                    val trimmed = query.trim()
                    if (trimmed.isEmpty()) {
                        _results.value = emptyList()
                        _isSearching.value = false
                    } else {
                        performSearch(trimmed)
                    }
                }
        }
    }

    private suspend fun performSearch(query: String) {
        _isSearching.value = true
        try {
            val inputMode =
                when (prefs.inputMode) {
                    "poj" -> ToneConverterModels.InputMode.POJ
                    else -> ToneConverterModels.InputMode.TL
                }

            val isCJK = query.any { it.code in 0x4E00..0x9FFF || it.code in 0x3400..0x4DBF || it.code in 0x20000..0x2A6DF }

            if (BuildConfig.DEBUG) {
                Log.d(TAG, "[SEARCH] query='$query' isCJK=$isCJK inputMode=$inputMode")
            }

            val searchResults =
                if (isCJK) {
                    root.lexicon.searchByHanzi(
                        input = query,
                        inputMode = inputMode,
                        limit = SEARCH_RESULT_LIMIT,
                    )
                } else {
                    root.lexicon.searchWithSources(
                        input = query,
                        inputMode = inputMode,
                        limit = SEARCH_RESULT_LIMIT,
                    )
                }

            if (BuildConfig.DEBUG) {
                Log.d(TAG, "[SEARCH] ${if (isCJK) "hanzi" else "roman"} path returned ${searchResults.size} results")
            }

            val customResults = searchCustomDictionary(query, isCJK)

            if (BuildConfig.DEBUG) {
                Log.d(TAG, "[SEARCH] custom dictionary returned ${customResults.size} results")
            }

            // Sort: KAUTIAN (教育部) first, then by frequency
            val sorted =
                searchResults.sortedWith(
                    compareByDescending<DictionarySearchResult> { DictionarySource.KAUTIAN in it.sources }
                        .thenByDescending { it.frequency },
                )
            val enabledSources = buildEnabledSources()
            val filtered =
                sorted.map { result ->
                    result.copy(sources = result.sources.filter { it in enabledSources })
                }
            _results.value = customResults + filtered
            _isSearching.value = false
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) {
                Log.e(TAG, "[SEARCH] Failed: ${e.message}", e)
            }
            _results.value = emptyList()
            _isSearching.value = false
        }
    }

    // Non-CJK custom dictionary search (matching iOS behavior)
    private suspend fun searchCustomDictionary(
        query: String,
        isCJK: Boolean,
    ): List<DictionarySearchResult> {
        if (isCJK) return emptyList()
        return try {
            val isToneAware = query.any { it.isDigit() }
            val searchPrefix =
                if (isToneAware) {
                    query.lowercase().replace("-", "").replace(" ", "")
                } else {
                    CustomDictionaryDerivation.generateNotone(query)
                }
            root.customDict
                .search(
                    prefix = searchPrefix,
                    isToneAware = isToneAware,
                    limit = SEARCH_RESULT_LIMIT,
                ).map { entry ->
                    DictionarySearchResult(
                        id = -2,
                        roman = entry.roman,
                        tl = entry.roman,
                        hanzi = entry.hanzi,
                        frequency = Int.MAX_VALUE,
                        sources = listOf(DictionarySource.CUSTOM),
                    )
                }
        } catch (e: Exception) {
            if (BuildConfig.DEBUG) {
                Log.w(TAG, "[SEARCH] Custom dictionary query failed: ${e.message}", e)
            }
            emptyList()
        }
    }
}
